rfcs/003-double-entry-ledger.md

```diff
@@ -0,0 +1,122 @@
+two rows in the `Transactions` table. One to add up to the Collective
+ledger (CREDIT) and the other to subtract from the User's ledger
+(DEBIT).
+
+However the transactions between the collective and the host, the
+platform and the payment processor (fees) don't have their own
+separate rows. The fees are stored as fields in both database rows but
+that don't quite describe the flow of funds between the collective and
+the institutions that are receiving the fees.
+
+This change will increase accuracy because it will simplify the way we
+store the transactions. It will also decrease duplicated code (likely
+reducing bugs) because it will provide a library that implements all
+the basic operations related to the ledger.
+
+## Goals
+
+1. Increase accuracy of the ledger
+   1. [ ] Create collective type "Payment Method"
+   2. [ ] Create separate transactions for fees
+2. Make it easier to build new features that depend on data from the
+   ledger
+   1. [ ] `libocledger`: wrap the most common operations related to
+           the ledger under a library that can be used across all the
+           places that interact with the ledger and decrease
+           duplicated code in the repository.
+
+### Create collective type "Payment Method"
+
+As mentioned before, the amount of the payment processor fee is stored
+in the field `paymentProcessorFeeInHostCurrency` in the `Transactions`
+table. To find out to which specific payment method the fee went to,
+the `Transactions` table also records the `PaymentMethodId`.
+
+This is different from how transactions are tracked between User &
+Collective. To address this difference, the Payment Method needs to
+have a ledger as well.
+
+### Create separate transactions for fees
+
+#### Use case example
+
+Given a transaction that represents moving $5 dollars from User to
+Collective with 5% of platform fee, 10% of host fee and 2.9%+30c of
+payment processor fee.
+
+##### How it is currently stored
+|            | User Ledger | Collective Ledger |
+|------------|------------:|------------------:|
+| Type       |       DEBIT |            CREDIT |
+| Amount     |       -4075 |              5000 |
+| Host Fee   |        -500 |              -500 |
+| Plat Fee   |        -250 |              -250 |
+| PP Fee     |        -175 |              -175 |
+| Net Amount |       -5000 |              4075 |
+
+##### How it should look like after this change
+
+|        |  User | Collective | Collective | Platform | Collective | Host | Collective | Payment Method |
+|--------|------:|-----------:|-----------:|---------:|-----------:|-----:|-----------:|---------------:|
+| Amount | -5000 |       5000 |       -250 |      250 |       -500 |  500 |       -175 |            175 |
+
+##### Where did the net amount go
+
+The net amount won't be a static value in the database anymore. The
+next section introduces an API that provides core operations to the
+ledger including calculate (& cache) the balance, getting fees etc.
```

Comment:
Generally like the idea that we don't need to store netAmount anywhere. Curious how you would want to implement it to make it easy to have this value everywhere we need it. 

Also, how would we validate that rows add up without a netAmount?

Reply:
That's something I started thinking about but again I'd certainly need more reading & thinking.

My initial thought was to implement that as a JavaScript function that retrieves all the transactions from a given `TransactionGroup` and a collective and returned the net amount. Something like this:

(I'm using the values present in the example of the API call `libledger.rows`)

```javascript
> libledger.netAmount('d3e36baa-69a2-4f6e-ac44-f93981e51c97', 58)
4325
> libledger.netAmount('d3e36baa-69a2-4f6e-ac44-f93981e51c97', 58)
-5000
```

This function would query the database and retrieve all the rows related to a given transaction. Since that can get expensive very quickly with a larger amount of transactions, I'd suggest creating a caching layer on top of these calls (initially thought of `redis`).